File: frontend/src/api/markets.ts
import client from './client';
import type {
  CreateMarketRequest,
  Market,
  MarketHistoryResponse,
  MarketsResponse,
  MarketWithPools,
  ProbabilitySnapshot,
  ResolveMarketRequest,
  UpdateMarketRequest,
} from '@/types';

interface FetchMarketsParams {
  category_id?: string;
  community_id?: string;
  status?: 'ouvert' | 'fermé';
  search?: string;
  sort?: 'volume' | 'newest';
  limit?: number;
  offset?: number;
}

export const marketsAPI = {
  fetchMarkets: async (params: FetchMarketsParams): Promise<MarketsResponse> => {
    const res = await client.get('/markets', { params });
    return res.data;
  },

  fetchMarket: async (id: string): Promise<MarketWithPools> => {
    const res = await client.get(`/markets/${id}`);
    return res.data;
  },

  createMarket: async (data: CreateMarketRequest): Promise<Market> => {
    const res = await client.post('/markets', data);
    return res.data;
  },

  updateMarket: async (id: string, data: UpdateMarketRequest): Promise<Market> => {
    const res = await client.patch(`/markets/${id}`, data);
    return res.data;
  },

  resolveMarket: async (id: string, winning_outcome_index: number): Promise<Market> => {
    const payload: ResolveMarketRequest = { winning_outcome_index };
    const res = await client.post(`/markets/${id}/resolve`, payload);
    return res.data;
  },

  fetchMarketHistory: async (id: string): Promise<ProbabilitySnapshot[]> => {
    const res = await client.get<MarketHistoryResponse>(`/markets/${id}/history`);
    return res.data.history;
  },

  fetchMarketBets: async (id: string): Promise<any[]> => {
    const res = await client.get(`/markets/${id}/bets`);
    return res.data.bets;
  },
};

File: crates/backend/src/routes/models/markets.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use sqlx::FromRow;
use uuid::Uuid;

#[derive(Debug, Serialize, Deserialize, Clone, FromRow)]
pub struct Market {
    pub id: Uuid,
    pub title: String,
    pub description: String,
    pub category_id: Uuid,
    pub community_id: Option<Uuid>,
    pub creator_id: Uuid,
    pub outcomes: Vec<String>,
    pub status: String,
    pub winning_outcome_index: Option<i32>,
    pub close_at: DateTime<Utc>,
    pub created_at: DateTime<Utc>,
}

#[derive(Debug, Serialize)]
pub struct MarketPool {
    pub outcome: String,
    pub total_points: i64,
    pub probability: f64,
}

#[derive(Debug, Serialize)]
pub struct UserPosition {
    pub outcome_index: i32,
    pub amount: i64,
}

#[derive(Debug, Serialize)]
pub struct MarketWithPools {
    #[serde(flatten)]
    pub market: Market,
    pub pools: Vec<MarketPool>,
    pub total_volume: i64,
    pub user_position: Option<UserPosition>,
}

#[derive(Debug, Serialize, Deserialize, Clone, FromRow)]
pub struct Bet {
    pub id: Uuid,
    pub market_id: Uuid,
    pub user_id: Uuid,
    pub outcome_index: i32,
    pub amount: i64,
    pub created_at: DateTime<Utc>,
}

#[derive(Debug, Serialize, Deserialize, Clone, FromRow)]
pub struct ProbabilitySnapshot {
    pub recorded_at: DateTime<Utc>,
    pub probabilities: Vec<f64>,
}

#[derive(Debug, Serialize)]
pub struct BetResponse {
    pub bet: Bet,
    pub new_balance: i64,
    pub probabilities_after: Vec<f64>,
}

#[derive(Debug, Serialize)]
pub struct MarketsResponse {
    pub markets: Vec<Market>,
    pub total: i64,
}

#[derive(Debug, Serialize)]
pub struct MarketHistoryResponse {
    pub history: Vec<ProbabilitySnapshot>,
}

#[derive(Debug, Deserialize)]
pub struct MarketListQuery {
    pub category_id: Option<Uuid>,
    pub community_id: Option<Uuid>,
    pub status: Option<String>,
    pub search: Option<String>,
    pub sort: Option<String>,
    pub limit: Option<i64>,
    pub offset: Option<i64>,
}

#[derive(Debug, Deserialize)]
pub struct CreateMarketRequest {
    pub title: String,
    pub description: String,
    pub category_id: Uuid,
    pub community_id: Option<Uuid>,
    pub outcomes: Vec<String>,
    pub close_at: DateTime<Utc>,
}

#[derive(Debug, Deserialize)]
pub struct UpdateMarketRequest {
    pub title: Option<String>,
    pub description: Option<String>,
    pub close_at: Option<DateTime<Utc>>,
}

#[derive(Debug, Deserialize)]
pub struct ResolveMarketRequest {
    pub winning_outcome_index: i32,
}

#[derive(Debug, Deserialize)]
pub struct PlaceBetRequest {
    pub outcome_index: i32,
    pub amount: i64,
}
